findDependencies.py.file

```diff
@@ -0,0 +1,193 @@
+    if not cache["uses"][file].has_key(x): cache["uses"][file][x] = {}
+    cache["usedby"][x][file] = 1
+    cache["uses"][file][x] = 1
+
+def pythonDeps(rel):
+  cache = {}
+  for root, dirs, files in os.walk("%s/src/" % rel):
+    for filename in files:
+      fpath = os.path.join(root, filename)
+      if re.search(r'.py$', fpath):
+        fname = fpath
+        file = fname
+        file = re.sub(r"^%s\/+src\/+" % rel, r'', file)
+        if not re.search(r'\/python\/', fname):
+          continue
+        with open(fname, 'r') as f:
+          for line in f.readlines():
+            if 'import ' in line:
+              line = line.rstrip('\n')
+              if re.search(r'^\s*#', line):
+                continue
+              match_from_import = re.search(r'^\s*from\s+([^\s]+)\s+import\s+', line)
+              match_import = re.search(r'^\s*import\s+([^\s]+)\s*', line)
+              if match_from_import:
+                createCache(match_from_import, cache, file)
+              elif match_import:
+                createCache(match_import, cache, file)
+  for type_ in ("uses","usedby"):
+    write2File("%s/etc/dependencies/py%s.out" % (rel, type_), cache, type_)
+
+def import2CMSSWDir(str, cache):
+  pyfiles = []
+  if not cache.has_key("pymodule"): cache["pymodule"] = {}
+  if not cache.has_key("noncmsmodule"): cache["noncmsmodule"] = {}
+  for s in str.split(","):
+    s = re.sub(r'\.', r'/', s)
+    if s in cache["pymodule"]:
+      pyfiles.append(cache["pymodule"][s])
+    elif s not in cache["noncmsmodule"]:
+      if os.path.exists("%s/python/%s.py" % (rel, s)):
+        match = re.search( r'^([^\/]+\/+[^\/]+)\/+(.+)$', s)
+        if match:
+          cache["pymodule"][s] = "%s/python/%s.py" % (match.group(1), match.group(2))
+          pyfiles.append("%s/python/%s.py" % (match.group(1), match.group(2)))
+      else: cache["noncmsmodule"][s] = 1
+  return pyfiles
+
+def data2json(infile):
+  jstr = ""
+  rebs=re.compile(',\s+"BuildSystem::[a-zA-Z0-9]+"\s+\)')
+  revar=re.compile("^\s*\$[a-zA-Z0-9]+\s*=")
+  reundef=re.compile('\s*undef,')
+  lines = [l.strip().replace(" bless(","").replace("'",'"').replace('=>',' : ')  for l in
+  gzip.open(infile).readlines()]
+  lines[0] = revar.sub("",lines[0])
+  lines[-1] = lines[-1].replace(";","")
+  for l in lines:
+    l =  reundef.sub(' "",',rebs.sub("", l.rstrip()))
+    jstr += l
+  return json.loads(jstr)
+
+def updateBFDeps(dir, pcache, cache):
+  bf = cache["dirs"][dir]
+  if not cache.has_key("uses"): cache["uses"] = {}
+  if bf in cache["uses"]: return 0
+  cache["uses"][bf] = {}
+  for pack in pcache["BUILDTREE"][dir]["RAWDATA"]["DEPENDENCIES"].keys():
+    if pack in cache["packs"]:
+      xdata = cache["packs"][pack]
+      updateBFDeps(xdata, pcache, cache)
+      xdata = cache["dirs"][xdata]
+      cache["uses"][bf][xdata] = 1
+      if not cache.has_key("usedby"): cache["usedby"] = {}
+      if not cache["usedby"].has_key(xdata): cache["usedby"][xdata] = {}
```

Comment:
you might add
usedby_cache = cache["usedby"] 
and then replace all the cache["used by"] with usedby_cache below

Reply:
There is no actual execution time improvement with this change (tried multiple times with "time" bash command)